master/internal/command/command.go

```diff
@@ -38,460 +37,245 @@ import (
+func (c *Command) startCmd(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	priorityChange := func(priority int) error {
+		return c.setNTSCPriority(priority, false)
+	}
+	if err := tasklist.GroupPriorityChangeRegistry.Add(c.jobID, priorityChange); err != nil {
+		return err
 	}
+	c.allocationID = model.AllocationID(fmt.Sprintf("%s.%d", c.taskID, 1))
 
-	results := []*command{}
-	for i := range snapshots {
-		cmd := commandFromSnapshot(ctx, pgDB, rm, &snapshots[i])
-		results = append(results, cmd)
+	if !c.restored {
+		if err := db.Bun().RunInTx(ctx, nil, c.registerJobAndTask); err != nil {
+			return err
+		}
+		if err := c.persistAndEvictContextDirectoryFromMemory(); err != nil {
+			return err
+		}
 	}
 
-	return results, nil
-}
+	priority := c.Config.Resources.Priority
+	if priority != nil {
+		if err := c.setNTSCPriority(*priority, true); err != nil {
+			return errors.Wrapf(err, "setting priority of task %v", c.taskID)
+		}
+	}
 
-func restoreCommandsByType(
-	ctx *actor.Context,
-	pgDB *db.PgDB,
-	rm rm.ResourceManager,
-	taskType model.TaskType,
-) error {
-	commands, err := remakeCommandsByType(ctx, pgDB, rm, taskType)
+	var idleWatcherConfig *sproto.IdleTimeoutConfig
+	if c.Config.IdleTimeout != nil && (c.WatchProxyIdleTimeout || c.WatchRunnerIdleTimeout) {
+		idleWatcherConfig = &sproto.IdleTimeoutConfig{
+			ServiceID:       string(c.taskID),
+			UseProxyState:   c.WatchProxyIdleTimeout,
+			UseRunnerState:  c.WatchRunnerIdleTimeout,
+			TimeoutDuration: time.Duration(*c.Config.IdleTimeout),
+			Debug:           c.Config.Debug,
+		}
+	}
+
+	err := task.DefaultService.StartAllocation(c.logCtx,
+		sproto.AllocateRequest{
+			AllocationID:        c.allocationID,
+			TaskID:              c.taskID,
+			JobID:               c.jobID,
+			JobSubmissionTime:   c.registeredTime,
+			IsUserVisible:       true,
+			Name:                c.Config.Description,
+			SlotsNeeded:         c.Config.Resources.Slots,
+			ResourcePool:        c.Config.Resources.ResourcePool,
+			FittingRequirements: sproto.FittingRequirements{SingleAgent: true},
+			ProxyPorts:          sproto.NewProxyPortConfig(c.GenericCommandSpec.ProxyPorts(), c.taskID),
+			IdleTimeout:         idleWatcherConfig,
+			Restore:             c.restored,
+			ProxyTLS:            c.TaskType == model.TaskTypeNotebook,
+		}, c.db, c.rm, c.GenericCommandSpec, c.onExit)
 	if err != nil {
 		return err
 	}
 
-	for _, cmd := range commands {
-		a, ok := ctx.ActorOf(cmd.taskID, cmd)
-		if !ok {
-			return fmt.Errorf("failed to recreate restored generic command actor %s", cmd.taskID)
-		}
+	// Once the command is persisted to the dbs & allocation starts, register it with the local job service.
+	jobservice.DefaultService.RegisterJob(c.jobID, c)
 
-		ctx.Ask(a, actor.Ping{}).Get()
-		ctx.Log().Debugf("restored generic command %s", cmd.taskID)
+	if err := c.persist(); err != nil {
+		c.syslog.WithError(err).Warnf("command persist failure")
```

Comment:
Should we return this error? And change to persist before StartAllocation? cc @stoksc

Reply:
+1. can change either way